xarray/core/dataarray.py

```diff
@@ -466,7 +466,17 @@ def _getitem_coord(self, key):
         return self._replace_maybe_drop_dims(var, name=key)
 
     def __getitem__(self, key):
-        if isinstance(key, basestring):
+        try:
+            is_coord_key = any([
+                isinstance(key, basestring),
+                key in set(self.dims).union(self.coords)
```

Comment:
I would hesitate to do this fix -- it means `array[2]` could pull out a coordinate if `2` is the name of a dimension. More generally, `DataArray.__getitem__` is only meant as a convenient short-cut for getting coordinates, which we might even want to deprecate entirely. `DataArray.coords.__getitem__` is the reliable way to get a coordinate out.

Instead, I think the better option would be to avoid making use of `**` unpacking below, in the expression `self.isel(**self._item_key_to_dict(key))`. This would be a little more involved, but I think the right fix would be to make most of these functions accept a positional dictionaries instead of `**kwargs`, as suggested in https://github.com/pydata/xarray/issues/1231.

Reply:
Yes, completely agree, thanks